c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -598,24 +786,14 @@ class DeviceCachingAllocator {
   }
 
   StatType get_stat_type_for_pool(const BlockPool& pool) {
-    if (&pool == &small_blocks) {
-      return StatType::SMALL_POOL;
-    } else if (&pool == &large_blocks) {
-      return StatType::LARGE_POOL;
-    } else {
-      TORCH_CHECK(false, "get_stat_type_for_pool: invalid pool");
-    }
+    return pool.is_small ? StatType::SMALL_POOL : StatType::LARGE_POOL;
   }
 
   bool should_split(const Block* block, size_t size) {
     size_t remaining = block->size - size;
-    if (block->pool == &small_blocks) {
-      return remaining >= kMinBlockSize;
-    } else if (block->pool == &large_blocks) {
-      return remaining > kSmallSize;
-    } else {
-      TORCH_CHECK(false, "should_split: invalid pool");
-    }
+    return (block->pool->is_small) ?
+           (remaining >= kMinBlockSize) :
+           (remaining > kSmallSize);
```

Comment:
are you sure invalid pool can't come here? There was a check for it before

Reply:
In an [earlier version (hopefully the link works)](https://github.com/pytorch/pytorch/pull/51436/files/987e4290b43264ce69ef20549a192a6b55401dec#diff-acc6337586bf9cdcf0a684380779300ec171897d05b8569bf439820dc8c93bd5R756-R764) I checked `block` against `&small_blocks`, `&large_blocks,` then a loop that checked against all the private pools' `&small_blocks` and `&large_blocks`, then finally an error if no match was found. It was kinda ugly and Ed said it was ["perf fishy"](https://github.com/pytorch/pytorch/pull/51436/files/987e4290b43264ce69ef20549a192a6b55401dec#r568191422). Using a new `pool->is_small` field made the code much simpler. If we still want to validate `block`, i have to restore all the original checks. I can do so if you want, but personally I don't think it's worth the ugliness. I have other checks that pools are present or not present when expected.

`get_stat_type_for_pool` is in the same boat.